functions/federated_aws_rp/utils.py

```diff
@@ -0,0 +1,298 @@
+import base64
+import json
+import logging
+import urllib.parse
+import http.cookies
+from typing import Optional
+
+from jose import jwt
+import requests
+import boto3
+
+from .config import CONFIG
+
+logger = logging.getLogger(__name__)
+
+
+class AccessDenied(Exception):
+    pass
+
+
+def base64_without_padding(data: bytes) -> str:
+    """Return the base64 encoding of data with padding stripped
+
+    https://tools.ietf.org/html/rfc7636#appendix-A
+    :param data: Bytes to encode
+    :return: A base64 encoded string with padding stripped
+    """
+    return base64.urlsafe_b64encode(data).decode("utf-8").rstrip("=")
+
+
+def encode_cookie_value(value: dict) -> str:
+    """Given a dictionary, JSON dump it and encode the result for a cookie
+
+    :param value: A dictionary to encode
+    :return: An encoded string to be set in a cookie
+    """
+    return http.cookies.SimpleCookie().value_encode(
+        json.dumps(value, separators=(',', ':')))[1]
+
+
+def get_store(cookie_header: str) -> dict:
+    """Given an HTTP cookie header return the store
+
+    Extract the cookie with the name CONFIG.cookie_name from the cookie_header
+    parse the JSON from it and return it
+
+    :param cookie_header: The content of the "Cookie" HTTP header
+    :return: The value of the cookie
+    """
+    cookies = http.cookies.SimpleCookie()
+    cookies.load(cookie_header)
+    if CONFIG.cookie_name in cookies:
+        try:
+            return json.loads(cookies[CONFIG.cookie_name].value)
+        except Exception:
+            logger.error('Unable to parse  JSON from "{}"'.format(
+                cookies[CONFIG.cookie_name].value
+            ))
+            return {}
+    else:
+        return {}
+
+
+def get_discovery_document(discovery_url: str) -> dict:
+    """Fetch the OIDC discovery document and enrich it with the JWKS
+
+    Fetch the the discovery_url and parse it into a dictionary. Fetch the
+    content of the jwks_uri and add it to the dictionary in the "jkws" key
+
+    :param discovery_url: The URL of the identity providers OIDC discovery url
+    :return: A dictionary containing information about the identity provider
+    """
+    result = requests.get(discovery_url).json()
+    result['jwks'] = requests.get(result['jwks_uri']).json()
```

Comment:
Just a note that this all has the same `requests` stuff we currently have with not handling exceptions.

Reply:
Good point. I think I should probably deal with this for all of the requests calls, do something graceful.

Reply:
I'll open an issue on this.